graphstorm-processing/graphstorm_processing/graph_loaders/dist_heterogeneous_loader.py

```diff
@@ -622,6 +628,121 @@ def csv_row(data: Row):
         )
         return filtered_key_list
 
+    @staticmethod
+    def _create_metadata_entry(path_list: Sequence[str]) -> dict:
+        return {
+            "format": {"name": FORMAT_NAME, "delimiter": DELIMITER},
+            "data": path_list,
+        }
+
+    def _write_pyarrow_table(
+        self, pa_table: pa.Table, out_path: str, num_files: int = 1
+    ) -> list[str]:
+        """Writes a PyArrow Table to multiple Parquet files.
+
+        Parameters
+        ----------
+        pa_table : pa.Table
+            PyArrow Table to write to storage.
+        out_path : str
+            Full filepath to write. Can be an S3 URI or local path.
+        num_files : int, optional
+            Number of files to write. Defaults to 1.
+
+        Returns
+        -------
+        list[str]
+            The list of files written, paths being relative to ``self.output_prefix``.
+        """
+        # Remove .parquet extension and add to file path instead
+        # makes the output consistent with self._write_df()
+        if out_path.endswith(".parquet"):
+            base_path = os.path.join(out_path[:-8], "parquet")
+        else:
+            base_path = os.path.join(out_path, "parquet")
+
+        if self.filesystem_type == FilesystemType.LOCAL:
+            os.makedirs(base_path, exist_ok=True)
+            pyarrow_fs = fs.LocalFileSystem()
+        else:
+            bucket, _ = s3_utils.extract_bucket_and_key(base_path)
+            region = s3_utils.get_bucket_region(bucket)
+            pyarrow_fs = fs.S3FileSystem(
+                region=region, retry_strategy=fs.AwsStandardS3RetryStrategy(max_attempts=10)
+            )
+            # When using S3FileSystem pyarrow expects bucket/key paths
+            if base_path.startswith("s3://"):
+                base_path = base_path[5:]
+
+        # Calculate rows per file
+        total_rows = len(pa_table)
+        rows_per_file = total_rows // num_files
+        remainder = total_rows % num_files
+
+        written_files = []
+        start_idx = 0
+
+        # Generate single UUIDs for all files
+        unique_id = uuid4()
+        for file_idx in range(num_files):
+            # Calculate end index for this file
+            rows_this_file = rows_per_file + (1 if file_idx < remainder else 0)
+            end_idx = start_idx + rows_this_file
+
+            # Slice the table
+            file_table = pa_table.slice(start_idx, rows_this_file)
+
+            # Generate Spark-style part filename
+            part_filename = os.path.join(
+                base_path,
+                f"part-{file_idx:05d}-{unique_id}.parquet",
```

Comment:
So unlike pyspark, pyarrow does not support the built-in naming?

Reply:
Yes, we could try using pyarrow.ds.dataset to do the writing to provide that but that actually complicates things. This allows us to have explicit distribution of rows.